### ISSUE TITLE
Make Poisson updates stable without a special `max_delta_step` default

### ISSUE THREAD

**Issue by RAMitchell** [dmlc/xgboost#12428]
The Poisson objective currently depends on `max_delta_step=0.7` in two places:

- The objective multiplies the Hessian by `exp(0.7)`.
- The tree code limits each leaf update to `0.7`.

The learner has special code to give both places the same value. This is hard to
follow and makes an internal stability detail into a user parameter.

The real Poisson gradient and Hessian are:

```text
gradient = mu - y
hessian  = mu
```

When `mu` is close to zero the update -g/h tends to infinity.

We can instead use:

```text
hessian = max(mu, y)
```

This uses the real Hessian when `mu >= y`. It only adds curvature when the
prediction is too small.

<img width="1476" height="864" alt="Image" src="https://github.com/user-attachments/assets/1c593e3e-def2-4904-a6b9-4438e3310d18" />

The proposed Hessian stays positive in the problem area and
returns to the real Hessian once `mu` reaches `y`.

## Why this is safe

For one leaf, let:

- `M` be the sum of the predicted means.
- `Y` be the sum of the labels.
- `H` be the sum of `max(mu, y)`.

Then `H` is at least both `M` and `Y`. The raw leaf update is therefore between
`-1` and `1`. For `learning_rate <= 1`, it moves toward the best Poisson leaf
value without passing it.

This is the smallest rowwise Hessian that gives this bound for every possible
leaf.

## Initial results

In a CPU histogram experiment with seven synthetic count problems and five
seeds:

- The proposal was stable in all 35 runs.
- Training Poisson loss never increased.
- The exact Hessian failed in all 10 deliberately bad-start runs.
- The proposal matched or improved the current method in six of seven cases.
- It reached its best range faster than the current method in every case.
- It was 0.46% worse in the remaining calibrated high-count case.

These tests used depth-4 trees, `learning_rate=0.3`, 20,000 training rows,
10,000 validation rows, and 100 boosting rounds.

**Comment by PrakshaaleJain** [comment]
Hi, this looks good to fix, if I may I would love to work on this issue.

**Comment by RAMitchell** [comment]
No sorry, this is just me outlining my notes and rationale. The change is trivial and I will do it myself.

**Comment by PrakshaaleJain** [comment]
Sure!

**Comment by RAMitchell** [comment]
An even better solution is a Halley update applied to the scalar leaf problem.

For a fixed Poisson leaf, define

$$
M=\sum_i w_i\mu_i,\qquad Y=\sum_i w_i y_i,\qquad R=Y/M.
$$

After adding leaf value $\Delta$, the loss is

$$
L(\Delta)=Me^\Delta-Y\Delta+C,
$$

so the exact leaf update is $\Delta^*=\log R$. Starting from $\Delta=0$:

$$
\Delta_{\mathrm{Newton}}=R-1,
$$

$$
\Delta_{\mathrm{majoriser}}
=\frac{Y-M}{K},
\qquad
K=\sum_i w_i\max(\mu_i,y_i),
$$

and applying Halley's method to the leaf score gives

$$
\Delta_{\mathrm{Halley}}
=\frac{2(Y-M)}{M+Y}
=\frac{2(R-1)}{R+1}.
$$

Although Halley normally uses the third derivative of the loss, Poisson has $L'''=L''$. No additional row statistic is therefore needed: using

$$
g_i=w_i(\mu_i-y_i),\qquad
\tilde h_i=\frac{w_i}{2}(\mu_i+y_i)
$$

makes the standard leaf calculation $-\sum g_i/\sum\tilde h_i$ produce the Halley update.

Halley is symmetric under $R\leftrightarrow1/R$, bounded by $(-2,2)$, and never crosses the exact optimum. Locally,

$$
\Delta_{\mathrm{Halley}}
=2\tanh\left(\frac{\log R}{2}\right)
=\log R-\frac{(\log R)^3}{12}+O((\log R)^5),
$$

so repeated full updates on a fixed leaf converge cubically.

<img width="2144" height="1130" alt="Image" src="https://github.com/user-attachments/assets/88be8a05-b520-4d3d-9504-049bd42b8a62" />

The majoriser is shown as a band because $K$ depends on the composition of the leaf:

$$
\max(M,Y)\le K\le M+Y.
$$

It is bounded by $[-1,1]$ and is more conservative. Halley therefore provides a middle ground: closer to the exact log-ratio update while remaining bounded, unlike Newton.

These statements assume $Y>0$, no regularization or clipping, and a full update $\eta=1$.
